File: backend/app/utils/plan_limits.py
"""Shared free vs Pro plan ceilings (server-side).

Keep in sync with ``src/lib/planLimits.ts``.
"""

from __future__ import annotations

from typing import Optional

FREE_LISTINGS_MAX_SIZE = 12
FREE_LISTINGS_MAX_PAGE = 1
FREE_BEST_PICKS_LIMIT = 6
FREE_ALERTS_LIMIT = 1
FREE_PRICE_DROPS_LIMIT = 3
FREE_SIMILAR_LIMIT = 4
FREE_PULSE_LIMIT = 6
FREE_TRENDS_MONTHS = 6
FREE_EV_MODELS_LIMIT = 3
FREE_VEHICLE_NEWS_LIMIT = 2
PRO_ALERTS_LIMIT = 20

# Full-access paid plans (not free-teased). Dealer includes Pro depth + yard tools.
PRO_PLANS = {"pro", "enterprise", "dealer"}


def is_free_browse_plan(plan: Optional[str], *, role: Optional[str] = None) -> bool:
    """True when the caller should receive free-tier teaser limits."""
    if str(role or "").strip().lower() == "admin":
        return False
    if plan is None:
        # Anonymous / open-mode callers are not free-teased here; app gate handles auth.
        return False
    return str(plan).strip().lower() not in PRO_PLANS


def take_last_months(rows: list, months: int) -> list:
    """Keep the newest N period-sorted rows for free teasers."""
    if not isinstance(rows, list) or months <= 0:
        return []
    if len(rows) <= months:
        return rows
    return rows[-months:]

File: src/lib/planLimits.ts
/**
 * Free vs Pro product limits.
 *
 * Free is a deliberate teaser: enough to feel Motormila’s value,
 * not enough to replace Pro for serious pricing / research work.
 * Admins and Pro/Enterprise subscribers are treated as full access.
 *
 * Product rule: free users can *open every page*, but each surface
 * soft-limits depth (counts, months, tabs, scores) instead of hard-blocking routes.
 */

export const FREE_LISTINGS_PAGE_SIZE = 12;
/** Free users only see page 1 of the market feed. */
export const FREE_MAX_LISTING_PAGES = 1;
/** Best Picks shortlist length on free. */
export const FREE_BEST_PICKS_LIMIT = 6;
/** Server + local alert cap for free. */
export const FREE_ALERTS_LIMIT = 1;
/** Official Pulse cards shown on free. */
export const FREE_PULSE_LIMIT = 6;
/** Trend / price-index history window (months) for free. */
export const FREE_TRENDS_MONTHS = 6;
/** Price-drop cards shown on free Best Picks / dashboard. */
export const FREE_PRICE_DROPS_LIMIT = 3;
/** Similar listings on listing detail for free. */
export const FREE_SIMILAR_LIMIT = 4;
/** Top EV models shown on free EV Hub. */
export const FREE_EV_MODELS_LIMIT = 3;
/** Calculator tabs free users can fully use. */
export const FREE_CALCULATOR_TABS = ["landed-cost"] as const;

export function hasFullPlatformAccess(opts: {
  hasProAccess: boolean;
  isAdmin?: boolean;
}): boolean {
  return Boolean(opts.hasProAccess || opts.isAdmin);
}

export function freeListingsVisibleTotal(marketTotal: number): number {
  return Math.min(Math.max(0, marketTotal), FREE_LISTINGS_PAGE_SIZE * FREE_MAX_LISTING_PAGES);
}

/** Keep the newest N months of a period-sorted series for free teasers. */
export function takeLastMonths<T>(rows: T[], months: number): T[] {
  if (!Array.isArray(rows) || months <= 0) return [];
  if (rows.length <= months) return rows;
  return rows.slice(rows.length - months);
}

export const freePlanCopy = {
  listingsTitle: "Free browse limit reached",
  listingsBody:
    "You’re seeing the first 12 live listings. Pro unlocks the full market feed, deal scores, and unlimited pagination.",
  picksTitle: "More elite deals on Pro",
  picksBody: "Free shows a short teaser shortlist. Pro unlocks the full ranked Best Picks board and price-cut feed.",
  alertsTitle: "One free alert included",
  alertsBody: "Pro unlocks deeper alert watches, richer match refresh, and WhatsApp notifications.",
  scoresTitle: "Deal scores are a Pro signal",
  scoresBody: "Upgrade to see fair-price scoring on every listing and sort by deal quality.",
  pulseTitle: "Limited Official Pulse on Free",
  pulseBody: "Free shows the latest 6 signals. Pro unlocks the full pulse desk and history.",
  trendsTitle: "6-month trend window on Free",
  trendsBody: "Pro unlocks full history plus district and condition filters.",
  calcTitle: "Starter calculator on Free",
  calcBody: "Landed-cost is open on Free. Lease, TCO, on-road fees, permits, and retention curves unlock on Pro.",
  indexTitle: "6-month index window on Free",
  indexBody: "Pro unlocks the full index history and segment breakdowns.",
  dealerTitle: "Dealer workspace teaser",
  dealerBody: "Claim and basic yard tools are open. Deep dealer intelligence unlocks on Pro / Dealer.",
  genericCta: "Upgrade to Pro",
} as const;
